src/mon/OSDMonitor.h

```diff
@@ -424,14 +424,18 @@ class OSDMonitor : public PaxosService {
   bool preprocess_beacon(MonOpRequestRef op);
   bool prepare_beacon(MonOpRequestRef op);
 
+  friend class C_UpdateCreatingPGs;
   using pgs_by_osd_epoch = std::map<int, std::map<epoch_t, std::set<pg_t>>>;
   pgs_by_osd_epoch creating_pgs_by_osd_epoch;
   pgs_by_osd_epoch pending_created_pgs;
+  epoch_t creating_pgs_epoch;	// the epoch when the pg mapping was calculated
   creating_pgs_t creating_pgs;
+  Spinlock creating_pgs_lock;
```

Comment:
all code block guarded by this lock is non-blocking code, so i think spin lock would suffice.